src/masternodes/mn_checks.cpp

```diff
@@ -2692,21 +2700,40 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
                         return Res::Err("Cannot payback this amount of loan for %s, either payback full amount or less than this amount!", loanToken->symbol);
             }
 
-            res = mnview.SubMintedTokens(loanToken->creationTx, subLoan);
-            if (!res)
-                return res;
+            if (height < Params().GetConsensus().FortCanningHillHeight || kv.first != DCT_ID{0})
+            {
+                res = mnview.SubMintedTokens(loanToken->creationTx, subLoan);
+                if (!res)
+                    return res;
+            }
 
             CalculateOwnerRewards(obj.from);
-            // subtract loan amount first, interest is burning below
-            res = mnview.SubBalance(obj.from, CTokenAmount{kv.first, subLoan});
-            if (!res)
-                return res;
 
-            // burn interest Token->USD->DFI->burnAddress
+            // sub amount from balance and burn interest Token->USD->DFI->burnAddress
             if (subInterest)
             {
-                LogPrint(BCLog::LOAN, "CLoanTakeLoanMessage(): Swapping %s interest to DFI - %lld, height - %d\n", loanToken->symbol, subInterest, height);
-                res = SwapToDFIOverUSD(mnview, kv.first, subInterest, obj.from, consensus.burnAddress, height);
+                if (height < Params().GetConsensus().FortCanningHillHeight || kv.first != DCT_ID{0})
+                {
+                    // subtract loan amount first, interest is burning below
+                    res = mnview.SubBalance(obj.from, CTokenAmount{tokenId, subLoan});
+                    if (!res)
+                        return res;
+
+                    LogPrint(BCLog::LOAN, "CLoanTakeLoanMessage(): Swapping %s interest to DFI - %lld, height - %d\n", loanToken->symbol, subInterest, height);
+                    res = SwapToDFIOverUSD(mnview, tokenId, subInterest, obj.from, consensus.burnAddress, height);
+                }
+                else
+                {
+                    CAmount subLoanInDFI = mnview.GetAmountInCurrency(subLoan, {"DFI","USD"}, false, true, true);
+                    res = TransferTokenBalance(DCT_ID{0}, subLoanInDFI, obj.from, consensus.burnAddress);
+                    if (!res)
+                        return res;
+
+                    LogPrint(BCLog::LOAN, "CLoanTakeLoanMessage(): Burning interest in DFI directly - %lld, height - %d\n", subInterest, height);
+                    CAmount subInterestInDFI = mnview.GetAmountInCurrency(subInterest, {"DFI","USD"}, false, true, true);
```

Comment:
This (in combination with the +1 in the reverse direction for `GetAmountInCurrency`) can lead to rounding errors where the available DFI balance is less than the amount that should be burned.

One possible fix would be defining `subInterestInDFI` as the minimum of `kv.second - subLoanInDFI` and the expression you have. I.e.,

    CAmount subInterestInDFI = min(kv.second - subLoanInDFI, mnview.GetAmountInCurrency(...) )

However, since when repaying with DFI, both the interest and the loan amount are burned. So this could be made simpler (and less prone to rounding issues) by combining both burns into one transaction? E.g.:

    CAmount subDebtInDFI = min(kv.second, mnview.GetAmountInCurrency(subInterest + subLoan, {"DFI","USD"}, false, true, true);
    res = TransferTokenBalance(DCT_ID{0}, subDebtInDFI, obj.from, consensus.burnAddress);

Reply:
I like the solution